### PR TITLE
Retire reference data instead of hard-deleting it (#289)

### LINKED ISSUE #289: Hard-deleting reference data silently corrupts all zero-based-id-as-index lookups (backend + frontend)
## Summary

The whole app resolves static reference data (items, item mods, skills, enemies, zones, challenges) **by array index**, relying on the documented "the Id of the entity is its index" invariant (see `docs/backend.md` → _Reference Data_, and the zero-based-Id seeding). That invariant only holds while ids stay **contiguous** (append-only). But the admin Content Authoring tools support **hard-deleting** any top-level reference record, and a hard delete of a non-terminal record creates a permanent id gap — after which **every index lookup for ids above the gap silently resolves to the wrong record**, on both the backend and the frontend, with no exception.

This is a latent defect (it requires an admin to delete a non-terminal reference record), but it is an unguarded sharp edge that silently corrupts gameplay data, so it is worth a deliberate decision.

## Why it happens

**Lookups are index-based** (a few representative sites):
- `Game.DataAccess/Repositories/Items.cs:46` — `LookupItem`: `items[itemId]` over an `OrderBy(i => i.Id)` list; `:51` — `GetItem`: `AllEntities()[itemId]`.
- `Game.DataAccess/Repositories/ItemMods.cs:51,56`, `Skills.cs:45`, `Enemies.cs:55`, `Zones.cs:45,52,58`, `Challenges.cs:37` — all the same `list[id]` shape.
- Frontend mirrors this: `staticData.items?.[id]` etc. in `UI/src/lib/battle/item.ts:16`, `item-mod.ts:7`, `routes/game/screens/challenges/challenges-view.svelte.ts:106,108,110,118,151`, `fight/boss/boss-view.svelte.ts:20`, `inventory/inventory-view.svelte.ts:183` (the frontend arrays are the `All()` projection, so they inherit the same ordering/gaps).

**Hard delete is exposed and reachable** — every admin reference repo has a top-level `delete` branch that calls `_entityStore.Delete(...)`:
- `AdminItems.cs:50`, `AdminItemMods.cs:47`, `AdminSkills.cs:38`, `AdminEnemies.cs:33`, `AdminZones.cs:77`, `AdminChallenges.cs:42`, `AdminTags.cs:29`.
- The Workbench surfaces this as a per-record Delete (see `docs/frontend.md` → _Admin Tools_; `EntityStore` `deleted` status).

**Concrete failure** — delete item id `5` out of `0..N`:
- The cache becomes `[0,1,2,3,4,6,7,…]`. `LookupItem(5)` returns the entity that was id `6`; `LookupItem(6)` returns id `7`; every id above the gap is off by one (cumulative per deletion). The original highest id becomes unreachable (returns `null`).
- Because `GetItem`/`GetItemMod`/`GetDomainEnemy`/`GetSkill`/`GetDomainZone` feed `BattleSnapshot` reconstruction and battle setup, a player's equipped item / applied mods / enemy loadout would resolve to the **wrong** reference object — corrupting player state and breaking frontend↔backend battle parity, silently.

## Possible directions (to evaluate, not prescriptive)

1. **Preserve the invariant — forbid gap-creating deletes.** Treat reference data as append-only: either remove the hard-delete affordance for top-level reference records, soft-delete with a tombstone that keeps the slot occupied, or only permit deleting the current highest id. Cheapest; keeps the index-as-id performance design intact. Probably needs the cache projection to keep a placeholder for a tombstoned slot so indices don't shift.
2. **Drop the index dependence — key lookups by id.** Back the caches with a `Dictionary<int,T>` (or `.find`/Map on the frontend) so a gap just yields "not found" instead of mis-resolution. Removes the fragility but gives up the documented Id-as-index optimization and touches every reference repo + the frontend lookup sites.
3. **Compact/renumber on delete** — rejected up front: it would rewrite ids that are referenced as FKs across the DB and as persisted ids in player state.

Whichever direction is chosen, the frontend should also be made consistent (see note below) and `docs/backend.md` _Reference Data_ should state the contiguity requirement explicitly.

## Related: inconsistent frontend lookup convention

Independently of the fix, the frontend currently **mixes** index access (`staticData.x[id]`, the sites above) with id-keyed `.find(e => e.id === id)` (`challenges-view.svelte.ts:92,96`, `attributes-view.svelte.ts:143`, `attribute-breakdown-view.svelte.ts:83,88`, `boss-view.svelte.ts:14`, the admin `reference.svelte.ts:119-123`) — sometimes in the **same file** (`boss-view.svelte.ts:14` uses `.find` for zones but `:20` indexes enemies by id). The `.find` form is the one that's robust to gaps. Standardizing on a single convention should fall out of whichever direction above is chosen.

## Acceptance

- A deliberate decision on the direction above, implemented across all reference repos + the frontend lookup sites.
- A test that deleting a non-terminal reference record does not mis-resolve a higher id (whichever guarantee the chosen direction provides).
- `docs/backend.md` _Reference Data_ updated to state the invariant/guarantee.

### PR DESCRIPTION
**Pull request by ginderjeremiah**
Closes #289.

## Problem

The whole app resolves static reference data (items, item mods, skills, enemies, zones, challenges) **by array index**, relying on the "the Id is its index" invariant. That only holds while ids stay contiguous. The admin Content Authoring tools exposed a **hard delete** of any top-level reference record, and deleting a non-terminal record opens a permanent id gap — after which every index lookup above the gap silently resolves to the *wrong* record, on both tiers, with no exception. It would also orphan persisted player-state ids and battle snapshots that reference the deleted record.

## Approach — Retire, not delete

Per discussion on the issue, reference records are now **retired** rather than deleted: a first-class content lifecycle state, not a generic soft-delete flag. A nullable `RetiredAt` timestamp (the same flag-plus-audit shape as the User `ArchivedAt`/`BannedAt` soft-delete) takes a record *out of circulation* while keeping it permanently valid for anything that already references it.

Guarantees:
- **The slot is permanent** — a retired record is never removed or renumbered, so the cache stays gap-free and index lookups can't mis-resolve. The contiguity invariant is now enforced *structurally* rather than by "only delete the highest id".
- **Still resolvable by id** — `GetItem`/`LookupItem`/`GetEnemy`/… still return a retired record, so a player's equipped item, a battle snapshot, or an authored boss/reward/unlock-gate FK keeps working. Retirement is a catalogue/authoring concern only — it is **not** on the lean `Game.Core` battle models, so a retired item a player owns fights identically.
- **Out of circulation** — a retired enemy is excluded from the random per-zone spawn tables. (There's no gameplay path that picks randomly from *all* reference data — items/mods/skills are authored challenge rewards, enemies/bosses are authored per-zone — so the circulation surface is small.)
- **No hard delete of a reference record** — `ChangeSetProcessor`'s delete handler is now optional; the six reference admin repos omit it and persist `RetiredAt` on edit, so a stray top-level `Delete` change throws instead of corrupting data. Child collections (attributes, mod slots, spawn rows, …) are still added/edited/deleted normally.

## Changes

**Backend**
- `RetiredAt` on the 6 reference entities + `AddReferenceDataRetiredAt` migration.
- Carried through the read contracts, the `ToContract` mappers, and the domain `Challenge` → `GetChallenges`.
- `ChangeSetProcessor.Apply` delete handler made optional (throws on a `Delete` when omitted); the 6 reference admin repos drop their hard-delete branch and persist `RetiredAt` on edit.
- `Enemies.BuildZoneEnemyTables` excludes retired enemies.

**Frontend**
- Regenerated API types (`retiredAt`).
- Workbench detail offers **Retire / Reinstate** (and **Remove** for a never-saved record) instead of Delete for reference entities; retired records show a muted "retired" badge and a dimmed list row. `EntityStore.setRetired`/`isRetired`; `retireable` flag on the 6 configs.

**Docs**
- `backend.md` → Reference Data rewritten with the retire guarantee + a "Retiring reference data" subsection.
- `game-design.md` → new "Content Lifecycle — Retiring Reference Data" section.

## Test plan

- [x] `GetEnemy_RetiringNonTerminalEnemy_DoesNotMisResolveHigherIds` — the issue's acceptance test: retiring a non-terminal record leaves every higher id resolving to itself.
- [x] `GetRandomEnemy_ExcludesRetiredEnemiesButKeepsThemResolvable` — spawn exclusion + still resolvable by id.
- [x] `AddEditEnemies_RetireEnemy_PersistsRetiredAtAndKeepsItResolvable` and `..._DeleteEnemy_IsRejectedAndLeavesRecordIntact` (API round-trip + delete rejection).
- [x] `ChangeSetProcessor` unit tests for the optional delete handler.
- [x] Frontend `entity-store` retire/reinstate tests (retire is a modify that preserves the slot, never a delete).
- [x] Full backend suite green (Core 337 / Infrastructure 26 / Api 290 / Application 108); frontend lint + `svelte-check` + 1051 unit tests green.

## Follow-ups (filed)

- #295 — exclude retired records from the admin authoring pickers.
- #296 — per-entity retirement edge cases (idling in a retired zone, in-progress challenge with a retired target/reward, retiring a zone/challenge gate, referenced-by warnings).
- #297 — standardize the frontend `.find`/index lookup convention and decide tag deletion semantics.

https://claude.ai/code/session_013rXj4DdWPhZCrSeLiWkbYG

---
_Generated by [Claude Code](https://claude.ai/code/session_013rXj4DdWPhZCrSeLiWkbYG)_

### CODE REVIEW

File: Game.DataAccess/Repositories/Enemies.cs [hunk 1]
```diff
@@ -91,10 +91,13 @@ public CoreEnemy GetRandomDomainEnemy(int zoneId, int level)
         /// Builds the per-zone enemy spawn tables from the in-memory enemy list. The ZoneEnemy weights are
         /// already eager-loaded by <see cref="AllEntities"/>, so no database query (and no lock) is required, and
         /// keying by zone id avoids the previous unbounded list growth for arbitrary or invalid ids.
+        /// Retired enemies are excluded so they no longer roll as random encounters, while still resolving
+        /// by id for existing references (e.g. an authored zone boss).
         /// </summary>
         private static Dictionary<int, ProbabilityTable<int>> BuildZoneEnemyTables(IEnumerable<Enemy> enemies)
         {
             return enemies
+                .Where(enemy => enemy.RetiredAt is null)
```
Comment (ginderjeremiah) [inline Game.DataAccess/Repositories/Enemies.cs:100]: Excluding retired enemies from the spawn tables is the right call, but it introduces a reachable runtime edge: if an admin retires **every** enemy assigned to an *active* (non-retired) zone, that zone drops out of `_zoneEnemyTables` entirely, and the next idle `GetRandomEnemy(zoneId)` throws `InvalidOperationException("Zone {zoneId} has no enemies to spawn.")` (line 70) — i.e. players idling there hit an exception on their next encounter.

It's not a regression (a hard delete of all the zone's enemies would have done the same), it's admin-only, and it overlaps the deferred per-entity edge cases in #296 — so this isn't a blocker. Worth either folding explicitly into #296's scope or adding a "last active enemy in a live zone" guard/warning on retire so it fails at authoring time rather than at spawn time.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ginderjeremiah — COMMENTED** [review]
## Review: Retire reference data instead of hard-deleting it (#289)

**Verdict:** Approve (intended) — posting as a `COMMENT` because I'm the PR author and GitHub won't take a gating verdict on my own PR. No blocking findings; one should-fix on frontend test coverage.

This is a clean, well-scoped solution to #289. Choosing **retire (soft-delete keeping the slot)** over re-keying lookups by id (issue direction 2) is the right trade-off: it preserves the documented Id-as-index optimization, enforces the contiguity invariant *structurally* rather than by "only delete the highest id" convention, and keeps retired records resolvable so persisted player state / battle snapshots / authored FKs stay valid. All three acceptance criteria are met (deliberate decision, a non-terminal-retire mis-resolution test, docs updated). Keeping `RetiredAt` off the lean `Game.Core` battle models is exactly right — a retired item a player owns fights identically, so there's no frontend/backend parity risk.

### Should-fix

- **Frontend component tests for the new retire UI are missing.** `WorkbenchDetail.svelte` gained a non-trivial three-way branch (Remove for a never-saved `id < 0` record → Reinstate for a retired one → Retire for an active one, gated on `entity.retireable`), and `WorkbenchList.svelte` gained the retired badge + dimmed row — but the only frontend test added is `entity-store.test.ts`. Both components already have established interaction suites (`WorkbenchDetail.test.ts` literally tests *"calls store.removeItem when the Delete button is clicked"* — the exact affordance this PR rewrote), so per the frontend testing rule ("every touched component with UI interaction has unit tests") the branch selection and badge rendering should be covered there. The store logic is well tested; it's the component wiring that's untested.

### Nits

- `Game.Core/Progress/Challenge.cs` now carries `RetiredAt`, an authoring-lifecycle field, on a gameplay domain model that doesn't read it (`UpdateChallengeProgress` ignores it). It's justified by the existing "`Challenge` is the single read+write contract" design (backend.md), so this is consistent rather than wrong — just noting the slight bleed in case the contract/domain split is ever revisited.
- The `Enemies.cs` all-retired-zone runtime edge — see the inline comment; fold into #296 or guard at authoring time.

### Test coverage

- **Backend — strong.** The issue's acceptance test (`GetEnemy_RetiringNonTerminalEnemy_DoesNotMisResolveHigherIds`), spawn exclusion + still-resolvable (`GetRandomEnemy_ExcludesRetiredEnemiesButKeepsThemResolvable`), the API round-trip (`AddEditEnemies_RetireEnemy_...`), delete rejection (`..._DeleteEnemy_IsRejectedAndLeavesRecordIntact`), and the `ChangeSetProcessor` unit tests for the now-optional delete handler all land the data-integrity-critical paths.
- **Frontend — store covered, components not.** `entity-store.test.ts` covers `isRetired`/`setRetired`/reinstate and asserts retire is a `modified` (slot-preserving) edit, never a delete. The component-level gap above is the one hole.

### What's good

- The optional-`delete`-handler change to `ChangeSetProcessor` is an elegant way to make a top-level hard-delete *fail loud* (throw) instead of silently corrupting — defense in depth alongside the UI no longer offering it.
- `RetiredAt` threaded consistently through entities → read contracts → mappers → codegen types; save payloads preserve it (spread for item/enemy/skill/item-mod, pass-through for challenge, explicit add for zone's destructured `toPrimaryDto`).
- Docs are genuinely good: `backend.md` Reference Data rewritten so the caveat now states the guarantee, and `game-design.md` gets a clear Content Lifecycle section. Both honestly scope what "out of circulation" means *today* (enemy spawns only) and defer the rest to #295/#296/#297.
- Tags correctly left on hard-delete (they're `.find`-keyed, not index-keyed) and split out to #297 rather than over-reaching.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Game.DataAccess/Repositories/Enemies.cs (hunk 1 above)
Comment (ginderjeremiah) [reply Game.DataAccess/Repositories/Enemies.cs:100]: Good catch. Since it's admin-only, not a regression, and fails loud (same as a zone with no enemies — already covered by `GetRandomEnemy_ValidZoneWithNoEnemies_ThrowsInvalidOperation`), I've folded it into #296 rather than guard it here, so the authoring-time warning lands with the other per-entity edge cases: https://github.com/ginderjeremiah/GameServer/issues/296#issuecomment

---
_Generated by [Claude Code](https://claude.ai/code)_
